plugins/continuous-toolbox/src/ContinuousFlyout.ts

```diff
@@ -0,0 +1,292 @@
+
+  /**
+   * Map from category name to its position in the flyout.
+   */
+  private scrollPositions = new Map<string, number>();
+
+  /**
+   * The percentage of the distance to the scrollTarget that should be
+   * scrolled at a time. Lower values will produce a smoother, slower scroll.
+   */
+  protected scrollAnimationFraction = 0.3;
+
+  /**
+   * Prevents the flyout from closing automatically when a block is dragged out.
+   */
+  override autoClose = false;
+
+  /**
+   * Creates a new ContinuousFlyout.
+   *
+   * @param workspaceOptions The injection options for the flyout's workspace.
+   */
+  constructor(workspaceOptions: Blockly.Options) {
+    super(workspaceOptions);
+
+    this.getWorkspace().setMetricsManager(
+      new ContinuousFlyoutMetrics(this.getWorkspace(), this),
+    );
+
+    this.getWorkspace().addChangeListener((e: Blockly.Events.Abstract) => {
+      if (e.type === Blockly.Events.VIEWPORT_CHANGE) {
+        this.selectCategoryByScrollPosition(-this.getWorkspace().scrollY);
+      }
+    });
+
+    this.setRecyclingEnabled(true);
+  }
+
+  /**
+   * Gets parent toolbox.
+   * Since we registered the ContinuousToolbox, we know that's its type.
+   *
+   * @returns Toolbox that owns this flyout.
+   */
+  private getParentToolbox(): ContinuousToolbox {
+    return this.targetWorkspace.getToolbox() as ContinuousToolbox;
+  }
+
+  /**
+   * Records scroll position for each category in the toolbox.
+   * The scroll position is determined by the coordinates of each category's
+   * label after the entire flyout has been rendered.
+   */
+  private recordScrollPositions() {
+    this.scrollPositions.clear();
+    this.getContents()
+      .filter(this.toolboxItemIsLabel.bind(this))
+      .map((item) => item.getElement())
+      .forEach((label) => {
+        this.scrollPositions.set(
+          label.getButtonText(),
+          label.getPosition().y - label.height,
+        );
+      });
+  }
+
+  /**
+   * Validates and typechecks that the given toolbox item represents a label.
+   *
+   * @param item The toolbox item to check.
+   * @returns True if the item represents a label in the flyout, and is a
+   *     Blockly.FlyoutButton.
+   */
+  protected toolboxItemIsLabel(
+    item: Blockly.FlyoutItem,
+  ): item is LabelFlyoutItem {
+    const element = item.getElement();
+    return !!(
+      item.getType() === 'label' &&
+      element instanceof Blockly.FlyoutButton &&
+      element.isLabel() &&
+      this.getParentToolbox().getCategoryByName(element.getButtonText())
+    );
+  }
+
+  /**
+   * Selects an item in the toolbox based on the scroll position of the flyout.
+   *
+   * @param position Current scroll position of the workspace.
+   */
+  private selectCategoryByScrollPosition(position: number) {
+    // If we are currently auto-scrolling, due to selecting a category by
+    // clicking on it, do not update the category selection.
+    if (this.scrollTarget) {
+      return;
+    }
```

Comment:
Better to keep these guard clauses to a single line:
```suggestion
    if (this.scrollTarget) return;
```

Reply:
Done